azure/services/aso/service_test.go

```diff
@@ -0,0 +1,386 @@
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[1], serviceName).Return(azure.NewOperationNotDoneError(&infrav1.Future{}))
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[2], serviceName).Return(nil)
+
+		s := &Service[*asoresourcesv1.ResourceGroup, *mock_aso.MockScope]{
+			Reconciler: reconciler,
+			Scope:      scope,
+			Specs:      specs,
+			name:       serviceName,
+		}
+
+		err := s.Delete(context.Background())
+		g.Expect(azure.IsOperationNotDoneError(err)).To(BeTrue())
+	})
+
+	t.Run("DeleteResource returns not done and another error", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+
+		mockCtrl := gomock.NewController(t)
+
+		scope := mock_aso.NewMockScope(mockCtrl)
+		specs := []azure.ASOResourceSpecGetter[*asoresourcesv1.ResourceGroup]{
+			mock_azure.NewMockASOResourceSpecGetter[*asoresourcesv1.ResourceGroup](mockCtrl),
+			mock_azure.NewMockASOResourceSpecGetter[*asoresourcesv1.ResourceGroup](mockCtrl),
+			mock_azure.NewMockASOResourceSpecGetter[*asoresourcesv1.ResourceGroup](mockCtrl),
+		}
+
+		reconciler := mock_aso.NewMockReconciler[*asoresourcesv1.ResourceGroup](mockCtrl)
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[0], serviceName).Return(azure.NewOperationNotDoneError(&infrav1.Future{}))
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[1], serviceName).Return(errors.New("non-not done error"))
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[2], serviceName).Return(azure.NewOperationNotDoneError(&infrav1.Future{}))
+
+		s := &Service[*asoresourcesv1.ResourceGroup, *mock_aso.MockScope]{
+			Reconciler: reconciler,
+			Scope:      scope,
+			Specs:      specs,
+			name:       serviceName,
+		}
+
+		err := s.Delete(context.Background())
+		g.Expect(err).To(MatchError("non-not done error"))
+	})
+
+	t.Run("CreateOrUpdateResource returns error and runs PostDeleteHook", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+
+		mockCtrl := gomock.NewController(t)
+
+		scope := mock_aso.NewMockScope(mockCtrl)
+		specs := []azure.ASOResourceSpecGetter[*asoresourcesv1.ResourceGroup]{
+			mock_azure.NewMockASOResourceSpecGetter[*asoresourcesv1.ResourceGroup](mockCtrl),
+		}
+
+		reconciler := mock_aso.NewMockReconciler[*asoresourcesv1.ResourceGroup](mockCtrl)
+		reconciler.EXPECT().DeleteResource(gomockinternal.AContext(), specs[0], serviceName).Return(errors.New("DeleteResource error"))
+
+		s := &Service[*asoresourcesv1.ResourceGroup, *mock_aso.MockScope]{
+			Reconciler: reconciler,
+			Scope:      scope,
+			Specs:      specs,
+			name:       serviceName,
+			PostDeleteHook: func(scopeParam *mock_aso.MockScope, err error) error {
+				g.Expect(scopeParam).To(BeIdenticalTo(scope))
+				g.Expect(err).To(MatchError("DeleteResource error"))
+				return errors.New("PostDeleteHook error")
+			},
+		}
+
+		err := s.Delete(context.Background())
+		g.Expect(err).To(MatchError("PostDeleteHook error"))
```

Comment:
These service unit tests are all well done and easy to follow. I like returning an error to validate that the hook funcs are called.